doc/sai-ptf/warmboot.md

```diff
@@ -0,0 +1,221 @@
+| --- | --- |
+| **Authors** | **Microsoft** |
+| **Status** | **In review** |
+| **Type** | **Richard Yu** |
+| **Created** | **22/03/2022** |
+| **Modified** | **22/03/2022** |
+| **SAI-Version** | **V1.7** |
+
+## Overriew
+
+### SAI-PTF for Warm reboot
+In order to use the SAI-PTF structure to verify the functionality in warm reboot scraniro, we need to add the following feature into the SAI-PTF structure
+1. Light weight docker which can expose the sai interface to invoke the sai interface remotely
+1. PTF test case can support differernt DUT running status with its different process - setup, runTest and teardown
+1. one test case can run in different DUT running status - before restart, starting and started
+1. Reuse already existing functional test cases and re-organize them into warm reboot structure
+
+## Saiserver container and test structure
+In order to test SAI interfaces, we need a light weight docker container which can help expose the SAI interfaces for a remote invocation.
+Test bed structure as below.
+### Test bed
+Those test will be run on the test bed structure as below, the components are:
+* PTF - running in a server which can connect to the target DUT
+* SAI server - running on a dut
+   ![Device_topology](img/Device_topology.jpg)
+*p.s. cause the sai testing will not depends on any sonic components, then there will be no specifical topology(T0 T1 T2) for testing.*
+### Test Structure
+---
+![Components](img/Component_topology.jpg)
+Test structure the chart above, components are:
+* PTF container - run test cases, and use a RPC client to invoke the SAI interfaces on DUT
+* SAI Server container - run inside DUT/switch, which exposes the SAI SDK APIs from the libsai
+* SAI-Qualify - Test controller, which is used to deploy and control the test running, meanwhile, manipulate the DUT on warm reboot.
+
+*For how to start a saiserver container please check the doc at
+[PTF-SAIv2 testing guide](https://github.com/Azure/sonic-mgmt/blob/master/docs/testbed/sai_quality/PTF-SAIv2TestingGuide.md) and 
+[Example: Start SaiServer Docker In DUT](https://github.com/Azure/sonic-mgmt/blob/master/docs/testbed/sai_quality/ExampleStartSaiServerDockerInDUT.md)*
+
+
+## PTF-SAI test supports DUT differernt running stage during warm reboot
+For PTF-SAI structure, it use three different method for three different steps in test
+- setUp, make settings
+- runTest, run test
+- tearDown, remove the setting and clear the test environment
+
+During a warm reboot, we need to split those three method in PTF-Test into three different stage
+- Before start, three methods
+  - setup, runTest, tearDown 
+- Starting, three methods
+  - setUp_starting, test_starting, tearDown_starting
+- After Starting, three methods.
+  - setUp_post_start, test_post_start, tearDown_post_start
+
+All those method extended by tag
+```
+@warm_setup
+@warm_test
+@warm_teardown
+```
+there is the sample code
+- Setup methods
+   ```python
+    @warm_setup
+    def setUp(self):
```

Comment:
Can you please list which of these 9 stages execute together at each of the 3 stages?

Reply:
do you mean this list?

- Before start, three methods
  - setup, runTest, tearDown 
- Starting, three methods
  - setUp_starting, test_starting, tearDown_starting
- After Starting, three methods.
  - setUp_post_start, test_post_start, tearDown_post_start